src/tcp/socks4.rs

```diff
@@ -0,0 +1,519 @@
+use crate::{Error, IntoTargetAddr, Result, TargetAddr, ToProxyAddrs};
+use futures_util::stream::{self, Fuse, Stream, StreamExt};
+use std::{
+    borrow::Borrow,
+    io,
+    net::{Ipv4Addr, SocketAddr},
+    ops::{Deref, DerefMut},
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tokio::{
+    io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt, ReadBuf},
+    net::TcpStream,
+};
+
+#[repr(u8)]
+#[derive(Clone, Copy)]
+enum CommandV4 {
+    Connect = 0x01,
+    Bind = 0x02,
+}
+
+/// A SOCKS4 client.
+///
+/// For convenience, it can be dereferenced to it's inner socket.
+#[derive(Debug)]
+pub struct Socks4Stream<S> {
+    socket: S,
+    target: TargetAddr<'static>,
+}
+
+impl<S> Deref for Socks4Stream<S> {
+    type Target = S;
+
+    fn deref(&self) -> &Self::Target {
+        &self.socket
+    }
+}
+
+impl<S> DerefMut for Socks4Stream<S> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.socket
+    }
+}
+
+impl Socks4Stream<TcpStream> {
+    /// Connects to a target server through a SOCKS4 proxy given the proxy
+    /// address.
+    ///
+    /// # Error
+    ///
+    /// It propagates the error that occurs in the conversion from `T` to
+    /// `TargetAddr`.
+    pub async fn connect<'t, P, T>(proxy: P, target: T) -> Result<Socks4Stream<TcpStream>>
+    where
+        P: ToProxyAddrs,
+        T: IntoTargetAddr<'t>,
+    {
+        Self::execute_command(proxy, target, None, CommandV4::Connect).await
+    }
+
+    /// Connects to a target server through a SOCKS5 proxy using given username,
```

Comment:
```suggestion
    /// Connects to a target server through a SOCKS4 proxy using given username,
```

Reply:
Same for a few other similar cases in the file

Reply:
Done